mathics/core/expression.py

```diff
@@ -881,17 +953,30 @@ def eval_range(indices):
                 new._leaves = tuple(dirty_leaves)
                 leaves = dirty_leaves
 
+        # If the attribute Flat is set, calls flatten with a callback
+        # that set leaves as unevaluated too.
+
         def flatten_callback(new_leaves, old):
             for leaf in new_leaves:
                 leaf.unevaluated = old.unevaluated
 
         if flat & attributes:
             new = new.flatten(new._head, callback=flatten_callback)
+
+        # If the attribute `Orderless` is set, sort the leaves, according to the
+        # `get_sort` criteria.
+        # the most expensive part of this is to build the sort key.
         if orderless & attributes:
             new.sort()
 
+        # Step 4:  Now, rebuilds the ExpressionCache, which tracks which symbols
```

Comment:
This seems to _create_ a new cache, not rebuild an existing one.

That said, right now I am not that concerned with the behavior because this is more a "nice to have" kind of thing rather than a necessary part of the operational logic.

Reply:
If the cache is rebuilt or created or updated depends on which changes the expression has suffered on the previous steps.